usd/src/usd_parser/USDData.cc

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#include "sdf/usd/usd_parser/USDData.hh"
+
+#include <string>
+#include <set>
+#include <unordered_map>
+
+// TODO(ahcorde):this is to remove deprecated "warnings" in usd, these warnings
+// are reported using #pragma message so normal diagnostic flags cannot remove
+// them. This workaround requires this block to be used whenever usd is
+// included.
+#pragma push_macro ("__DEPRECATED")
+#undef __DEPRECATED
+#include <pxr/usd/usd/primCompositionQuery.h>
+#include <pxr/usd/usd/primRange.h>
+#include <pxr/usd/usdPhysics/scene.h>
+#include <pxr/usd/usdShade/material.h>
+#pragma pop_macro ("__DEPRECATED")
+
+#include <ignition/common/Filesystem.hh>
+#include <ignition/common/Util.hh>
+
+#include "sdf/usd/usd_parser/utils.hh"
+#include "sdf/Material.hh"
+
+namespace sdf {
+// Inline bracke to help doxygen filtering.
+inline namespace SDF_VERSION_NAMESPACE {
+//
+namespace usd {
+  /// \brief USDStage private data.
+  class USDData::Implementation
+  {
+    public:
+      /// File name of the main stage
+      std::string filename;
+
+      /// Directory where the main stage is located
+      std::string directoryPath;
+
+      /// map with the filename and its own Stage data
+      std::unordered_map<std::string, std::shared_ptr<USDStage>> references;
+
+      /// Models
+      std::set<std::string> models;
+
+      /// Materials availables in the stage and substages
+      std::unordered_map<std::string, sdf::Material> materials;
+
+      /// Add all subdirectories that are inside the stage folder
+      /// This will help us to find other stages and/or textures.
+      /// \param[in] _path Path of the subdirectory to add
+      void AddSubdirectories(const std::string &_path)
```

Comment:
I'm not very sure if we should manage usd references manually, I think stage traversal automatically traverse referenced layers and prims. Stage loading is also very complicated, references can be made using different file formats, using different protocols and across different filesystems, we are only handling the most basic case of all referenced stage being in the same filesystem and being in the same directory hierachy.

I can try and see if I can find usd apis for locating, reading files etc so that we don't need this.